File: tests/ten_runtime/integration/go/close_app_cmd_go/close_app_cmd_go_app/ten_packages/extension/default_extension_go/extension.go
//
// Copyright © 2025 Agora
// This file is part of TEN Framework, an open source project.
// Licensed under the Apache License, Version 2.0, with certain conditions.
// Refer to the "LICENSE" file in the root directory for more information.
//

package default_extension_go

import (
	ten "ten_framework/ten_runtime"
)

type defaultExtension struct {
	ten.DefaultExtension
}

func (ext *defaultExtension) OnConfigure(tenEnv ten.TenEnv) {
	tenEnv.Log(ten.LogLevelDebug, "OnConfigure")
	tenEnv.OnConfigureDone()
}

func (ext *defaultExtension) OnCmd(
	tenEnv ten.TenEnv,
	cmd ten.Cmd,
) {
	cmdName, _ := cmd.GetName()
	if cmdName == "close" {
		// Testing get/set property to/from json with empty path.
		testCmd, _ := ten.NewCmd("test")
		testCmd.SetPropertyFromJSONBytes("", []byte("{\"key\":\"value\"}"))

		testValue, _ := testCmd.GetPropertyToJSONBytes("")
		tenEnv.Log(ten.LogLevelDebug, "testValue: "+string(testValue))
		// Testing end.

		closeAppCmd, _ := ten.NewCmd("ten:close_app")

		err := closeAppCmd.SetDests(ten.Loc{
			AppURI:        ten.Ptr(""),
			GraphID:       ten.Ptr(""),
			ExtensionName: ten.Ptr(""),
		})
		if err != nil {
			tenEnv.Log(ten.LogLevelError, "Failed to SetDests:"+err.Error())
			return
		}

		err = tenEnv.SendCmd(closeAppCmd, nil)
		if err != nil {
			tenEnv.Log(
				ten.LogLevelError,
				"Failed to send close cmd:"+err.Error(),
			)
			return
		}

		cmdResult, _ := ten.NewCmdResult(ten.StatusCodeOk, cmd)
		cmdResult.SetPropertyString("detail", "ok")

		tenEnv.ReturnResult(cmdResult, nil)
	}
}

func newAExtension(name string) ten.Extension {
	return &defaultExtension{}
}

func init() {
	// Register addon.
	err := ten.RegisterAddonAsExtension(
		"default_extension_go",
		ten.NewDefaultExtensionAddon(newAExtension),
	)
	if err != nil {
		panic("Failed to register addon.")
	}
}

File: tests/ten_runtime/integration/python/async_io_basic_python/async_io_basic_python_app/ten_packages/extension/default_extension_python/extension.py
#
# Copyright © 2025 Agora
# This file is part of TEN Framework, an open source project.
# Licensed under the Apache License, Version 2.0, with certain conditions.
# Refer to the "LICENSE" file in the root directory for more information.
#
import threading
import asyncio
from ten_runtime import (
    Extension,
    TenEnv,
    Cmd,
    CmdResult,
    LogLevel,
)


class DefaultExtension(Extension):
    async def __thread_routine(self, ten_env: TenEnv):
        ten_env.log(LogLevel.INFO, "__thread_routine start")

        self.loop = asyncio.get_running_loop()

        # Mock async operation before on_start_done
        await asyncio.sleep(1)

        ten_env.on_start_done()

        # Suspend the thread until stopEvent is set
        await self.stopEvent.wait()

    async def stop_thread(self):
        self.stopEvent.set()

    async def send_cmd_async(self, ten_env: TenEnv, cmd: Cmd) -> CmdResult:
        ten_env.log(LogLevel.INFO, "send_cmd_async")
        q = asyncio.Queue(maxsize=10)
        ten_env.send_cmd(
            cmd,
            lambda ten_env, result, error: asyncio.run_coroutine_threadsafe(
                q.put([result, error]), self.loop
            ),  # type: ignore
        )

        [result, error] = await q.get()
        if error is not None:
            raise Exception(error.error_message())
        return result

    def __init__(self, name: str) -> None:
        super().__init__(name)
        self.name = name
        self.stopEvent = asyncio.Event()

    def on_init(self, ten_env: TenEnv) -> None:
        ten_env.on_init_done()

    def on_start(self, ten_env: TenEnv) -> None:
        ten_env.log(LogLevel.DEBUG, "on_start")

        self.thread = threading.Thread(
            target=asyncio.run, args=(self.__thread_routine(ten_env),)
        )

        # Then 'on_start_done' will be called in the thread
        self.thread.start()

    def on_deinit(self, ten_env: TenEnv) -> None:
        ten_env.log(LogLevel.INFO, "on_deinit")
        ten_env.on_deinit_done()

    async def on_cmd_async(self, ten_env: TenEnv, cmd: Cmd) -> None:
        ten_env.log(LogLevel.INFO, "on_cmd_async")

        cmd_json, _ = cmd.get_property_to_json()
        ten_env.log(LogLevel.INFO, "on_cmd_async json: " + cmd_json)

        # Mock async operation, e.g. network, file I/O
        await asyncio.sleep(1)

        # Send a new command to other extensions and wait for the result. The
        # result will be returned to the original sender.
        new_cmd = Cmd.create("hello")
        cmd_result = await self.send_cmd_async(ten_env, new_cmd)

        cmd_result_json, _ = cmd_result.get_property_to_json()

        new_result = CmdResult.create(cmd_result.get_status_code(), cmd)
        new_result.set_property_from_json(None, cmd_result_json)

        ten_env.return_result(new_result)

    def on_cmd(self, ten_env: TenEnv, cmd: Cmd) -> None:
        ten_env.log(LogLevel.INFO, "on_cmd")

        asyncio.run_coroutine_threadsafe(
            self.on_cmd_async(ten_env, cmd), self.loop
        )

    def on_stop(self, ten_env: TenEnv) -> None:
        ten_env.log(LogLevel.INFO, "on_stop")

        if self.thread.is_alive():
            asyncio.run_coroutine_threadsafe(self.stop_thread(), self.loop)
            self.thread.join()

        ten_env.on_stop_done()
